<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Machine learning pitfalls â€¢ SIAMCAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Machine learning pitfalls">
<meta property="og:description" content="SIAMCAT">
<meta property="og:image" content="/logo.png">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><script type="text/javascript">
  var _paq = _paq || [];
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u="//tr-denbi.embl.de/heimdall/";
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '1']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SIAMCAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SIAMCAT_vignette.html">Get Started</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/SIAMCAT_vignette.html">Introduction to SIAMCAT</a>
    </li>
    <li>
      <a href="../articles/SIAMCAT_confounder.html">SIAMCAT example with confounders</a>
    </li>
    <li>
      <a href="../articles/SIAMCAT_holdout.html">Holdout testing with SIAMCAT</a>
    </li>
    <li>
      <a href="../articles/SIAMCAT_meta.html">SIAMCAT machine learning meta-analysis</a>
    </li>
    <li>
      <a href="../articles/SIAMCAT_ml_pitfalls.html">Machine learning pitfalls</a>
    </li>
    <li>
      <a href="../articles/SIAMCAT_read-in.html">SIAMCAT input formats</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://twitter.com/siamcat_dev">
    <span class="fas fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/zellerlab/siamcat">
    <span class="fas fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="SIAMCAT_ml_pitfalls_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Machine learning pitfalls</h1>
                        <h4 data-toc-skip class="author">Jakob Wirbel and Georg Zeller</h4>
            <address class="author_afil">
      EMBL Heidelberg<br><a class="author_email" href="mailto:#"></a><a href="mailto:georg.zeller@embl.de" class="email">georg.zeller@embl.de</a>
      </address>
                  
            <h4 data-toc-skip class="date">Date last modified: 2020-11-07</h4>
      
      
      <div class="hidden name"><code>SIAMCAT_ml_pitfalls.Rmd</code></div>

    </div>

    
    
<div id="about-this-vignette" class="section level1">
<h1 class="hasAnchor">
<a href="#about-this-vignette" class="anchor"></a>About This Vignette</h1>
<p>In this vignette, we want to explore two pitfalls for machine learning analysis that can lead to overly optimistic performance estimates.<br>
When setting up cross-validation workflows, the main objective is usually to estimate how well a trained model would perform on external data, which is specifically important when considering biomarker discovery. However, more complex workflows involving feature selection or time-course data can be challenging to setup correctly. Incorrect workflows in which information leaks from the test to the training data can lead to overfitting and poor generalization to external datasets.<br>
Here, we focus on supervised feature selection and the naive splitting of dependent data.</p>
<div id="setup" class="section level2">
<h2 class="hasAnchor">
<a href="#setup" class="anchor"></a>Setup</h2>
<p>First, we load the packages needed to perform the analyses.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="http://tidyverse.tidyverse.org">"tidyverse"</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st">"SIAMCAT"</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="supervised-feature-selection" class="section level1">
<h1 class="hasAnchor">
<a href="#supervised-feature-selection" class="anchor"></a>Supervised Feature Selection</h1>
<p>Supervised feature selection means that the label information is taken into account before the cross-validation split. Within this procedure, the features are selected if they are associated with the label (for example after differential abundance testing), using the complete dataset for the calculation of feature association and leaving no data aside for unbiased model evaluation.<br>
A correct way to perform feature selection would be to nest the selection step into the cross-validation procedure. That means that the calculation of feature association is performed for each training fold separately.</p>
<div id="load-the-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-data" class="anchor"></a>Load the Data</h2>
<p>As an example, we are going to use two datasets of colorectal cancer (CRC) which are available through the <code>curatedMetagenomicData</code> package.<br>
Since the model trainig procedure takes a long time, this vignette is not evaluated upon build of the package, but if you execute the code chunks for yourself, you should get similar results.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/waldronlab/curatedMetagenomicData">"curatedMetagenomicData"</a></span><span class="op">)</span></code></pre></div>
<p>First, we are going to load the dataset from <a href="https://doi.org/10.1038/s41591-019-0405-7">Thomas et al</a> as training dataset.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="st">'ThomasAM_2018a.metaphlan_bugs_list.stool'</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/curatedMetagenomicData.html">curatedMetagenomicData</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, dryrun<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="va">feat.t</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">assayData</span><span class="op">$</span><span class="va">exprs</span>
<span class="co"># clean up metaphlan profiles to contain only species-level abundances</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="va">feat.t</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span>, pattern<span class="op">=</span><span class="st">'s__'</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="va">feat.t</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span>,pattern<span class="op">=</span><span class="st">'t__'</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>As an external dataset, we are going to use the data from <a href="https://doi.org/10.15252/msb.20145645">Zeller et al.</a>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="st">'ZellerG_2014.metaphlan_bugs_list.stool'</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/curatedMetagenomicData/man/curatedMetagenomicData.html">curatedMetagenomicData</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">x</span>, dryrun<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="va">feat.z</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">@</span><span class="va">assayData</span><span class="op">$</span><span class="va">exprs</span>
<span class="co"># clean up metaphlan profiles to contain only species-level abundances</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="va">feat.z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span>, pattern<span class="op">=</span><span class="st">'s__'</span><span class="op">)</span>,<span class="op">]</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="va">feat.z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span>x<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span>,pattern<span class="op">=</span><span class="st">'t__'</span>, invert <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,<span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colSums.html">colSums</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span> <span class="op">!=</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html">t</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span><span class="op">/</span><span class="fl">100</span><span class="op">)</span></code></pre></div>
<p>We can also extract the corresponding metadata from the <code>combined_metadata</code> object which is part of the <code>curatedMetagenomicData</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta.t</span> <span class="op">&lt;-</span> <span class="va">combined_metadata</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">dataset_name</span> <span class="op">==</span> <span class="st">'ThomasAM_2018a'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">study_condition</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'control'</span>, <span class="st">'CRC'</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">meta.t</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta.t</span><span class="op">$</span><span class="va">sampleID</span>
<span class="va">meta.z</span> <span class="op">&lt;-</span> <span class="va">combined_metadata</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">dataset_name</span> <span class="op">==</span> <span class="st">'ZellerG_2014'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">study_condition</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'control'</span>, <span class="st">'CRC'</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">meta.z</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta.z</span><span class="op">$</span><span class="va">sampleID</span></code></pre></div>
<p>The MetaPhlAn2 profiler used for the profiles outputs only species which are present in the dataset. Therefore, we can have the case that there are species in the matrix for <code>ThomasAM_2018</code> which are not present in the matrix for <code>ZellerG_2014</code> and vice verse. In order to use them as training and external test set for <code>SIAMCAT</code>, we have to first make sure that the set of features for both datasets overlap completely (see also the <strong>Holdout Testing with SIAMCAT</strong> vignette).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">species.union</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">union</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span><span class="op">)</span>
<span class="co"># add Zeller_2014-only species to the Thomas_2018 matrix</span>
<span class="va">add.species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span><span class="op">(</span><span class="va">species.union</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span><span class="op">)</span>
<span class="va">feat.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">feat.t</span>, 
                <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">add.species</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span>,
                       dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">add.species</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">feat.t</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co"># add Thomas_2018-only species to the Zeller_2014 matrix</span>
<span class="va">add.species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span><span class="op">(</span><span class="va">species.union</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span><span class="op">)</span>
<span class="va">feat.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">feat.z</span>, 
                <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span><span class="op">(</span><span class="fl">0</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">add.species</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span>,
                       dimnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">add.species</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">feat.z</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Now, we are ready to start the model training process. For this, we chose three different feature selection cutoffs and prepare a tibble to hold the results:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">fs.cutoff</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">100</span>, <span class="fl">250</span><span class="op">)</span>

<span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, 
                    study.test<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/double.html">double</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="train-model-without-feature-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#train-model-without-feature-selection" class="anchor"></a>Train Model without Feature Selection</h2>
<p>First, we will train a model without any feature selection, using all the available features. We add it to the results matrix twice (both with <code>correct</code> and <code>incorrect</code>) for easier plotting later.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.t</span>, meta<span class="op">=</span><span class="va">meta.t</span>,
                    label<span class="op">=</span><span class="st">'study_condition'</span>, case<span class="op">=</span><span class="st">'CRC'</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter.features.html">filter.features</a></span><span class="op">(</span><span class="va">sc.obj.t</span>, filter.method <span class="op">=</span> <span class="st">'prevalence'</span>,
                            cutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.features.html">normalize.features</a></span><span class="op">(</span><span class="va">sc.obj.t</span>,
                               norm.method <span class="op">=</span> <span class="st">'log.std'</span>,
                               norm.param<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>log.n0<span class="op">=</span><span class="fl">1e-05</span>,
                                               sd.min.q<span class="op">=</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.data.split.html">create.data.split</a></span><span class="op">(</span><span class="va">sc.obj.t</span>,
                              num.folds <span class="op">=</span> <span class="fl">10</span>, num.resample <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/train.model.html">train.model</a></span><span class="op">(</span><span class="va">sc.obj.t</span>, method<span class="op">=</span><span class="st">'lasso'</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">)</span>
<span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">)</span>

<span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="st">'full'</span>, type<span class="op">=</span><span class="st">'correct'</span>, 
          study.test<span class="op">=</span><span class="st">'Thomas_2018'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="st">'full'</span>, type<span class="op">=</span><span class="st">'incorrect'</span>, 
          study.test<span class="op">=</span><span class="st">'Thomas_2018'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p>We then also apply the model to the external dataset and record the generalization to another dataset:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.z</span>, meta<span class="op">=</span><span class="va">meta.z</span>,
                    label<span class="op">=</span><span class="st">'study_condition'</span>, case<span class="op">=</span><span class="st">'CRC'</span><span class="op">)</span>
<span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t</span>, <span class="va">sc.obj.z</span><span class="op">)</span>
<span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">)</span>
<span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="st">'full'</span>, type<span class="op">=</span><span class="st">'correct'</span>, 
          study.test<span class="op">=</span><span class="st">'Zeller_2014'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="st">'full'</span>, type<span class="op">=</span><span class="st">'incorrect'</span>, 
          study.test<span class="op">=</span><span class="st">'Zeller_2014'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
</div>
<div id="incorrect-procedure-train-with-supervised-feature-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#incorrect-procedure-train-with-supervised-feature-selection" class="anchor"></a>Incorrect Procedure: Train with Supervised Feature Selection</h2>
<p>For the incorrect feature selection procedure, we can test the features for differential abundance, using the complete dataset, and then chose the top associated features.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc.obj.t</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check.associations.html">check.associations</a></span><span class="op">(</span><span class="va">sc.obj.t</span>, detect.lim <span class="op">=</span> <span class="fl">1e-05</span>,
                               fn.plot <span class="op">=</span> <span class="st">'assoc_plot.pdf'</span><span class="op">)</span>
<span class="va">mat.assoc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/associations.html">associations</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">)</span>
<span class="va">mat.assoc</span><span class="op">$</span><span class="va">species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">mat.assoc</span><span class="op">)</span>
<span class="co"># sort by p-value</span>
<span class="va">mat.assoc</span> <span class="op">&lt;-</span> <span class="va">mat.assoc</span> <span class="op">%&gt;%</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">arrange</span><span class="op">(</span><span class="va">p.val</span><span class="op">)</span></code></pre></div>
<p>Based on the P values from the <code>check.association</code> function, we now chose <code>X</code> number of features on which to train the model.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="va">fs.cutoff</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># select x number of features based on p-value ranking</span>
  <span class="va">feat.train.red</span> <span class="op">&lt;-</span> <span class="va">feat.t</span><span class="op">[</span><span class="va">mat.assoc</span> <span class="op">%&gt;%</span>
                             <span class="fu">slice</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq_len</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
                             <span class="fu">pull</span><span class="op">(</span><span class="va">species</span><span class="op">)</span>,<span class="op">]</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.train.red</span>, meta<span class="op">=</span><span class="va">meta.t</span>,
                         label<span class="op">=</span><span class="st">'study_condition'</span>, case<span class="op">=</span><span class="st">'CRC'</span><span class="op">)</span>
  <span class="co"># normalize the features without filtering</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.features.html">normalize.features</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span>,
                                    norm.method <span class="op">=</span> <span class="st">'log.std'</span>,
                                    norm.param<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>log.n0<span class="op">=</span><span class="fl">1e-05</span>,
                                                    sd.min.q<span class="op">=</span><span class="fl">0</span><span class="op">)</span>,
                                    feature.type <span class="op">=</span> <span class="st">'original'</span><span class="op">)</span>
  <span class="co"># take the same cross validation split as before</span>
  <span class="fu"><a href="../reference/data_split.html">data_split</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/data_split.html">data_split</a></span><span class="op">(</span><span class="va">sc.obj.t</span><span class="op">)</span>
  <span class="co"># train</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/train.model.html">train.model</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span>, method <span class="op">=</span> <span class="st">'lasso'</span><span class="op">)</span>
  <span class="co"># make predictions</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">)</span>
  <span class="co"># evaluate predictions and record the result</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, type<span class="op">=</span><span class="st">'incorrect'</span>, 
            study.test<span class="op">=</span><span class="st">'Thomas_2018'</span>,
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># apply to the external dataset and record the result</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.z</span>, meta<span class="op">=</span><span class="va">meta.z</span>,
                    label<span class="op">=</span><span class="st">'study_condition'</span>, case<span class="op">=</span><span class="st">'CRC'</span><span class="op">)</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span>, <span class="va">sc.obj.z</span><span class="op">)</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, type<span class="op">=</span><span class="st">'incorrect'</span>, 
            study.test<span class="op">=</span><span class="st">'Zeller_2014'</span>, 
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="correct-procedure-train-with-nested-feature-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#correct-procedure-train-with-nested-feature-selection" class="anchor"></a>Correct Procedure: Train with Nested Feature Selection</h2>
<p>Feature selection can be performed correctly if it is nested within the cross-validation procedure. We can do it using <code>SIAMCAT</code> by specifying the <code>perform.fs</code> parameter in the <code>train.model</code> function.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span> <span class="op">(</span><span class="va">x</span> <span class="kw">in</span> <span class="va">fs.cutoff</span><span class="op">)</span><span class="op">{</span>
  <span class="co"># train using the original SIAMCAT object </span>
  <span class="co"># with correct version of feature selection</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/train.model.html">train.model</a></span><span class="op">(</span><span class="va">sc.obj.t</span>, method <span class="op">=</span> <span class="st">'lasso'</span>, 
                             perform.fs <span class="op">=</span> <span class="cn">TRUE</span>,
                             param.fs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>thres.fs <span class="op">=</span> <span class="va">x</span>,
                                             method.fs <span class="op">=</span> <span class="st">"AUC"</span>,
                                             direction<span class="op">=</span><span class="st">'absolute'</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># make predictions</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">)</span>
  <span class="co"># evaluate predictions and record the result</span>
  <span class="va">sc.obj.t.fs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, type<span class="op">=</span><span class="st">'correct'</span>, 
            study.test<span class="op">=</span><span class="st">'Thomas_2018'</span>,
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># apply to the external dataset and record the result</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.z</span>, meta<span class="op">=</span><span class="va">meta.z</span>,
                    label<span class="op">=</span><span class="st">'study_condition'</span>, case<span class="op">=</span><span class="st">'CRC'</span><span class="op">)</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.t.fs</span>, <span class="va">sc.obj.z</span><span class="op">)</span>
  <span class="va">sc.obj.z</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, type<span class="op">=</span><span class="st">'correct'</span>, 
            study.test<span class="op">=</span><span class="st">'Zeller_2014'</span>, 
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">sc.obj.z</span><span class="op">@</span><span class="va">eval_data</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="plot-the-results" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-the-results" class="anchor"></a>Plot the Results</h2>
<p>Now, we can plot the resulting performance estimates for the cross-validation and the external validation as well:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">auroc.all</span> <span class="op">%&gt;%</span>
  <span class="co"># facetting for plotting</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>split<span class="op">=</span><span class="fu">case_when</span><span class="op">(</span><span class="va">study.test</span><span class="op">==</span><span class="st">"Thomas_2018"</span><span class="op">~</span>
                           <span class="st">'Cross validation (Thomas 2018)'</span>,
                         <span class="cn">TRUE</span><span class="op">~</span><span class="st">"External validation (Zeller 2014)"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># convert to factor to enforce ordering</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>cutoff<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">cutoff</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">fs.cutoff</span>, <span class="st">'full'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">cutoff</span>, y<span class="op">=</span><span class="va">AUC</span>, col<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">geom_line</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>group<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span><span class="va">split</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_y_continuous</span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span>, expand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">'Features selected'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">'AUROC'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">scale_colour_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'correct'</span><span class="op">=</span><span class="st">'blue'</span>, <span class="st">'incorrect'</span><span class="op">=</span><span class="st">'red'</span><span class="op">)</span>,
                        name<span class="op">=</span><span class="st">'Feature selection procedure'</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">theme</span><span class="op">(</span>panel.grid.minor <span class="op">=</span> <span class="fu">element_blank</span><span class="op">(</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">'bottom'</span><span class="op">)</span></code></pre></div>
<p><img src="SIAMCAT_ml_pitfalls_files/figure-html/plot_auroc-1.png" width="700"></p>
<p>As you can see, the incorrect feature selection procedure leads to inflated AUROC values but lower generalization to a truly external dataset, especially when very few features were selected. In contrast, the correct procedure gives a lower cross-validation results but a better estimation for how the model would perform on external data.</p>
</div>
</div>
<div id="naive-splitting-of-dependent-data" class="section level1">
<h1 class="hasAnchor">
<a href="#naive-splitting-of-dependent-data" class="anchor"></a>Naive Splitting of Dependent Data</h1>
<p>Another issue in machine learning workflows can occur when samples are not independent. For example, microbiome samples taken from the same individual at different time points are usually more similar to each other than to samples from other individuals. If these samples are split randomly in a naive cross-validation procedure, the case could arise that samples from the same individual will end up in the training and the test fold. In this case, the model would learn to generalize across time-points for the same individual compared to the desired model that should learn to distinguish the label across individuals.<br>
To avoid this issue, dependent measurements need to be blocked during cross-validation, to ensure that samples within the same block will stay in the same fold (for training and testing).</p>
<div id="load-the-data-1" class="section level2">
<h2 class="hasAnchor">
<a href="#load-the-data-1" class="anchor"></a>Load the Data</h2>
<p>As an example, we are going to use several datasets of Crohnâ€™s disease (CD) which are available through the EMBL cluster. The data have already been filtered and cleaned.<br>
Since the model training would take again quite a long time, this part of the vignette is not evaluated upon building of the package, but you should be able to execute it yourself.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data.location</span> <span class="op">&lt;-</span> <span class="st">'https://www.embl.de/download/zeller/'</span>

<span class="co"># metadata</span>
<span class="va">meta.all</span> <span class="op">&lt;-</span> <span class="fu">read_tsv</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data.location</span>, <span class="st">'CD_meta/meta_all.tsv'</span><span class="op">)</span><span class="op">)</span>
<span class="co">## </span>
<span class="co">## â”€â”€ Column specification â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€</span>
<span class="co">## cols(</span>
<span class="co">##   Sample_ID = col_character(),</span>
<span class="co">##   Group = col_character(),</span>
<span class="co">##   Library_Size = col_double(),</span>
<span class="co">##   Individual_ID = col_character(),</span>
<span class="co">##   Timepoint = col_double(),</span>
<span class="co">##   Study = col_character()</span>
<span class="co">## )</span>

<span class="co"># features</span>
<span class="va">feat.motus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">data.location</span>, <span class="st">'CD_meta/feat_rel_filt.tsv'</span><span class="op">)</span>,
                         sep<span class="op">=</span><span class="st">'\t'</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span>,
                         check.names <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>When we look at the number of samples and number of individuals, we see that that there are several samples per individual for example in the <code>HMP2</code> study.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x</span> <span class="op">&lt;-</span> <span class="va">meta.all</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Study</span>, <span class="va">Group</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">summarise</span><span class="op">(</span>n.all<span class="op">=</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, .groups<span class="op">=</span><span class="st">'drop'</span><span class="op">)</span>
<span class="va">y</span> <span class="op">&lt;-</span> <span class="va">meta.all</span> <span class="op">%&gt;%</span> 
  <span class="fu">select</span><span class="op">(</span><span class="va">Study</span>, <span class="va">Group</span>, <span class="va">Individual_ID</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Study</span>, <span class="va">Group</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">summarize</span><span class="op">(</span>n.indi<span class="op">=</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>,  .groups<span class="op">=</span><span class="st">'drop'</span><span class="op">)</span>
<span class="fu">full_join</span><span class="op">(</span><span class="va">x</span>,<span class="va">y</span><span class="op">)</span>
<span class="co">## Joining, by = c("Study", "Group")</span>
<span class="co">## # A tibble: 10 x 4</span>
<span class="co">##    Study         Group n.all n.indi</span>
<span class="co">##    &lt;chr&gt;         &lt;chr&gt; &lt;int&gt;  &lt;int&gt;</span>
<span class="co">##  1 Franzosa_2019 CD       88     88</span>
<span class="co">##  2 Franzosa_2019 CTR      56     56</span>
<span class="co">##  3 He_2017       CD       49     49</span>
<span class="co">##  4 He_2017       CTR      53     53</span>
<span class="co">##  5 HMP2          CD      583     50</span>
<span class="co">##  6 HMP2          CTR     357     26</span>
<span class="co">##  7 Lewis_2015    CD      294     85</span>
<span class="co">##  8 Lewis_2015    CTR      25     25</span>
<span class="co">##  9 metaHIT       CD       21     13</span>
<span class="co">## 10 metaHIT       CTR      71     59</span></code></pre></div>
<p>Therefore, we are going to train a model on the <code>HMP2</code> study. However, the number of samples per individual varies quite a lot across samples, therefore we want to randomly select a set of 5 samples per individual:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta.all</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Study</span><span class="op">==</span><span class="st">'HMP2'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Individual_ID</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">summarise</span><span class="op">(</span>n<span class="op">=</span><span class="fu">n</span><span class="op">(</span><span class="op">)</span>, .groups<span class="op">=</span><span class="st">'drop'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">pull</span><span class="op">(</span><span class="va">n</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html">hist</a></span><span class="op">(</span><span class="fl">20</span><span class="op">)</span></code></pre></div>
<p><img src="SIAMCAT_ml_pitfalls_files/figure-html/hmp_samples-1.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># sample 5 samples per individual</span>
<span class="va">meta.train</span> <span class="op">&lt;-</span> <span class="va">meta.all</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Study</span><span class="op">==</span><span class="st">'HMP2'</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Individual_ID</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">sample_n</span><span class="op">(</span><span class="fl">5</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu">distinct</span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">meta.train</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta.train</span><span class="op">$</span><span class="va">Sample_ID</span></code></pre></div>
<p>For evaluation, we only want a single sample per individual, therefore we can create a new matrix removing repeated samples for the other studies:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">meta.ind</span> <span class="op">&lt;-</span> <span class="va">meta.all</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Individual_ID</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Timepoint</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">Timepoint</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ungroup</span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p>Lastly, we can already create a tibble to hold the resulting AUROC values:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="op">(</span>type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, study.test<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">character</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>,
                    AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/double.html">double</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="train-with-naive-cross-validation" class="section level2">
<h2 class="hasAnchor">
<a href="#train-with-naive-cross-validation" class="anchor"></a>Train with Naive Cross-validation</h2>
<p>The naive way to split samples for cross-validation does not take into account the dependency between samples. Therefore, the pipeline would look basically like this:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.motus</span>, meta<span class="op">=</span><span class="va">meta.train</span>,
                        label<span class="op">=</span><span class="st">'Group'</span>, case<span class="op">=</span><span class="st">'CD'</span><span class="op">)</span>
<span class="va">sc.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/normalize.features.html">normalize.features</a></span><span class="op">(</span><span class="va">sc.obj</span>, norm.method <span class="op">=</span> <span class="st">'log.std'</span>,
                             norm.param<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>log.n0<span class="op">=</span><span class="fl">1e-05</span>,
                                             sd.min.q<span class="op">=</span><span class="fl">1</span><span class="op">)</span>,
                             feature.type <span class="op">=</span> <span class="st">'original'</span><span class="op">)</span>
<span class="va">sc.obj.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.data.split.html">create.data.split</a></span><span class="op">(</span><span class="va">sc.obj</span>, num.folds <span class="op">=</span> <span class="fl">10</span>, num.resample <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>
<span class="va">sc.obj.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/train.model.html">train.model</a></span><span class="op">(</span><span class="va">sc.obj.naive</span>, method<span class="op">=</span><span class="st">'lasso'</span><span class="op">)</span>
<span class="va">sc.obj.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.naive</span><span class="op">)</span>
<span class="va">sc.obj.naive</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.naive</span><span class="op">)</span>
<span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>type<span class="op">=</span><span class="st">'naive'</span>, study.test<span class="op">=</span><span class="st">'HMP2'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/eval_data.html">eval_data</a></span><span class="op">(</span><span class="va">sc.obj.naive</span><span class="op">)</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="train-with-blocked-cross-validation" class="section level2">
<h2 class="hasAnchor">
<a href="#train-with-blocked-cross-validation" class="anchor"></a>Train with Blocked Cross-validation</h2>
<p>The correct way to to take into account repeated samples would be to block the cross-validation procedure by individuals. This way, samples from the same individual will always end up in the same fold. This can be performed in <code>SIAMCAT</code> by specifying the <code>inseparable</code> parameter in the <code>create.data.split</code> function:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sc.obj.block</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create.data.split.html">create.data.split</a></span><span class="op">(</span><span class="va">sc.obj</span>, num.folds <span class="op">=</span> <span class="fl">10</span>, num.resample <span class="op">=</span> <span class="fl">10</span>,
                                  inseparable <span class="op">=</span> <span class="st">'Individual_ID'</span><span class="op">)</span>
<span class="va">sc.obj.block</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/train.model.html">train.model</a></span><span class="op">(</span><span class="va">sc.obj.block</span>, method<span class="op">=</span><span class="st">'lasso'</span><span class="op">)</span>
<span class="va">sc.obj.block</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.block</span><span class="op">)</span>
<span class="va">sc.obj.block</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.block</span><span class="op">)</span>
<span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
  <span class="fu">add_row</span><span class="op">(</span>type<span class="op">=</span><span class="st">'blocked'</span>, study.test<span class="op">=</span><span class="st">'HMP2'</span>, 
          AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/eval_data.html">eval_data</a></span><span class="op">(</span><span class="va">sc.obj.block</span><span class="op">)</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="apply-to-external-datasets" class="section level2">
<h2 class="hasAnchor">
<a href="#apply-to-external-datasets" class="anchor"></a>Apply to External Datasets</h2>
<p>Now, we can apply both models to external datasets and record the resulting accuracy:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/sets.html">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">meta.all</span><span class="op">$</span><span class="va">Study</span><span class="op">)</span>, <span class="st">'HMP2'</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">meta.test</span> <span class="op">&lt;-</span> <span class="va">meta.ind</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span><span class="op">(</span><span class="va">Study</span><span class="op">==</span><span class="va">i</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">meta.test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta.test</span><span class="op">$</span><span class="va">Sample_ID</span>
  <span class="co"># apply naive model</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.motus</span>, meta<span class="op">=</span><span class="va">meta.test</span>, 
                         label<span class="op">=</span><span class="st">'Group'</span>, case<span class="op">=</span><span class="st">'CD'</span><span class="op">)</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.naive</span>, <span class="va">sc.obj.test</span><span class="op">)</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.test</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>type<span class="op">=</span><span class="st">'naive'</span>, study.test<span class="op">=</span><span class="va">i</span>,
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/eval_data.html">eval_data</a></span><span class="op">(</span><span class="va">sc.obj.test</span><span class="op">)</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
  <span class="co"># apply blocked model</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/siamcat.html">siamcat</a></span><span class="op">(</span>feat<span class="op">=</span><span class="va">feat.motus</span>, meta<span class="op">=</span><span class="va">meta.test</span>, 
                         label<span class="op">=</span><span class="st">'Group'</span>, case<span class="op">=</span><span class="st">'CD'</span><span class="op">)</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make.predictions.html">make.predictions</a></span><span class="op">(</span><span class="va">sc.obj.block</span>, <span class="va">sc.obj.test</span><span class="op">)</span>
  <span class="va">sc.obj.test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/evaluate.predictions.html">evaluate.predictions</a></span><span class="op">(</span><span class="va">sc.obj.test</span><span class="op">)</span>
  <span class="va">auroc.all</span> <span class="op">&lt;-</span> <span class="va">auroc.all</span> <span class="op">%&gt;%</span> 
    <span class="fu">add_row</span><span class="op">(</span>type<span class="op">=</span><span class="st">'blocked'</span>, study.test<span class="op">=</span><span class="va">i</span>,
            AUC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/eval_data.html">eval_data</a></span><span class="op">(</span><span class="va">sc.obj.test</span><span class="op">)</span><span class="op">$</span><span class="va">auroc</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div id="plot-the-results-1" class="section level2">
<h2 class="hasAnchor">
<a href="#plot-the-results-1" class="anchor"></a>Plot the Results</h2>
<p>Now, we can compare the resulting AUROC values between the two different approaches:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">auroc.all</span> <span class="op">%&gt;%</span>
  <span class="co"># convert to factor to enforce ordering</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">type</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'naive'</span>, <span class="st">'blocked'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="co"># facetting for plotting</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>CV<span class="op">=</span><span class="fu">case_when</span><span class="op">(</span><span class="va">study.test</span><span class="op">==</span><span class="st">'HMP2'</span><span class="op">~</span><span class="st">'CV'</span>, <span class="cn">TRUE</span><span class="op">~</span><span class="st">'External validation'</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x<span class="op">=</span><span class="va">study.test</span>, y<span class="op">=</span><span class="va">AUC</span>, fill<span class="op">=</span><span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_bar</span><span class="op">(</span>stat<span class="op">=</span><span class="st">'identity'</span>, position <span class="op">=</span> <span class="fu">position_dodge</span><span class="op">(</span><span class="op">)</span>, col<span class="op">=</span><span class="st">'black'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">coord_cartesian</span><span class="op">(</span>ylim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'red'</span>, <span class="st">'blue'</span><span class="op">)</span>, name<span class="op">=</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span><span class="op">~</span><span class="va">CV</span>, space <span class="op">=</span> <span class="st">'free'</span>, scales <span class="op">=</span> <span class="st">'free'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">''</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ylab</span><span class="op">(</span><span class="st">'AUROC'</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme</span><span class="op">(</span>legend.position <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.8</span>, <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="SIAMCAT_ml_pitfalls_files/figure-html/plot_results_db-1.png" width="700"></p>
<p>As you can see, the naive cross-validation procedure leads to a inflated performance estimation compared to the blocked cross-validation. However, when assessing generalization to truly external datasets, the blocked procedure results in better performance.</p>
</div>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">## R version 4.0.2 (2020-06-22)</span>
<span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">## Running under: macOS Catalina 10.15.5</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRblas.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] SIAMCAT_1.11.1    phyloseq_1.32.0   mlr_2.18.0        ParamHelpers_1.14</span>
<span class="co">##  [5] forcats_0.5.1     stringr_1.4.0     dplyr_1.0.4       purrr_0.3.4      </span>
<span class="co">##  [9] readr_1.4.0       tidyr_1.1.2       tibble_3.0.6      ggplot2_3.3.3    </span>
<span class="co">## [13] tidyverse_1.3.0   BiocStyle_2.16.1 </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] colorspace_2.0-0    ellipsis_0.3.1      rprojroot_2.0.2    </span>
<span class="co">##   [4] XVector_0.28.0      fs_1.5.0            rstudioapi_0.13    </span>
<span class="co">##   [7] farver_2.0.3        fansi_0.4.2         lubridate_1.7.9.2  </span>
<span class="co">##  [10] xml2_1.3.2          PRROC_1.3.1         codetools_0.2-18   </span>
<span class="co">##  [13] splines_4.0.2       cachem_1.0.4        knitr_1.31         </span>
<span class="co">##  [16] ade4_1.7-16         jsonlite_1.7.2      pROC_1.17.0.1      </span>
<span class="co">##  [19] gridBase_0.4-7      broom_0.7.4         cluster_2.1.1      </span>
<span class="co">##  [22] dbplyr_2.1.0        BiocManager_1.30.10 compiler_4.0.2     </span>
<span class="co">##  [25] httr_1.4.2          backports_1.2.1     assertthat_0.2.1   </span>
<span class="co">##  [28] Matrix_1.3-2        fastmap_1.1.0       cli_2.3.0          </span>
<span class="co">##  [31] htmltools_0.5.1.1   prettyunits_1.1.1   tools_4.0.2        </span>
<span class="co">##  [34] igraph_1.2.6        gtable_0.3.0        glue_1.4.2         </span>
<span class="co">##  [37] LiblineaR_2.10-8    reshape2_1.4.4      fastmatch_1.1-0    </span>
<span class="co">##  [40] Rcpp_1.0.6          parallelMap_1.5.0   Biobase_2.48.0     </span>
<span class="co">##  [43] cellranger_1.1.0    pkgdown_1.6.1.9000  vctrs_0.3.6        </span>
<span class="co">##  [46] Biostrings_2.56.0   multtest_2.44.0     ape_5.4-1          </span>
<span class="co">##  [49] nlme_3.1-152        iterators_1.0.13    xfun_0.21          </span>
<span class="co">##  [52] rvest_0.3.6         lifecycle_0.2.0     beanplot_1.2       </span>
<span class="co">##  [55] zlibbioc_1.34.0     MASS_7.3-53.1       scales_1.1.1       </span>
<span class="co">##  [58] ragg_0.4.1          hms_1.0.0           parallel_4.0.2     </span>
<span class="co">##  [61] biomformat_1.16.0   rhdf5_2.32.4        RColorBrewer_1.1-2 </span>
<span class="co">##  [64] BBmisc_1.11         curl_4.3            yaml_2.2.1         </span>
<span class="co">##  [67] gridExtra_2.3       memoise_2.0.0       stringi_1.5.3      </span>
<span class="co">##  [70] highr_0.8           S4Vectors_0.26.1    desc_1.2.0         </span>
<span class="co">##  [73] corrplot_0.84       foreach_1.5.1       checkmate_2.0.0    </span>
<span class="co">##  [76] permute_0.9-5       BiocGenerics_0.34.0 shape_1.4.5        </span>
<span class="co">##  [79] matrixStats_0.58.0  rlang_0.4.10        pkgconfig_2.0.3    </span>
<span class="co">##  [82] systemfonts_1.0.1   evaluate_0.14       lattice_0.20-41    </span>
<span class="co">##  [85] Rhdf5lib_1.10.1     labeling_0.4.2      tidyselect_1.1.0   </span>
<span class="co">##  [88] plyr_1.8.6          magrittr_2.0.1      bookdown_0.21      </span>
<span class="co">##  [91] R6_2.5.0            IRanges_2.22.2      generics_0.1.0     </span>
<span class="co">##  [94] DBI_1.1.1           pillar_1.4.7        haven_2.3.1        </span>
<span class="co">##  [97] withr_2.4.1         mgcv_1.8-33         survival_3.2-7     </span>
<span class="co">## [100] modelr_0.1.8        crayon_1.4.1        utf8_1.1.4         </span>
<span class="co">## [103] rmarkdown_2.6       progress_1.2.2      grid_4.0.2         </span>
<span class="co">## [106] readxl_1.3.1        data.table_1.13.6   vegan_2.5-7        </span>
<span class="co">## [109] infotheo_1.2.0      reprex_1.0.0        digest_0.6.27      </span>
<span class="co">## [112] textshaping_0.3.0   glmnet_4.1          stats4_4.0.2       </span>
<span class="co">## [115] munsell_0.5.0</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Konrad Zych, Jakob Wirbel, Georg Zeller.</p>
</div>

<div class="privacy">
  <p><strong>Privacy:</strong> Usage of this site follows
      <a href="https://www.embl.de/aboutus/privacy_policy/index.html">EMBLâ€™s Privacy Policy</a><br>
      In accordance with that policy, we use Matomo to collect anonymised data on visits to, downloads from, and searches of this site.
      Contact <a href="mailto:zeller@embl.de">Georg Zeller</a> for details.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
